L1Trigger/Phase2L1GMT/plugins/Phase2L1TGMTSAMuonProducer.cc

```diff
@@ -0,0 +1,196 @@
+
+Phase2L1TGMTSAMuonProducer::~Phase2L1TGMTSAMuonProducer() {}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void Phase2L1TGMTSAMuonProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+  edm::Handle<l1t::MuonBxCollection> muon;
+  iEvent.getByToken(muonToken_, muon);
+
+  // Output
+  std::vector<SAMuon> prompt;
+  std::vector<SAMuon> displaced;
+
+  for (int bx = muon->getFirstBX(); bx <= muon->getLastBX(); ++bx) {
+    //TODO: We are expecting to send all BX. Using bx0 for now
+    if (bx != 0) {
+      continue;
+    }
+
+    for (uint i = 0; i < muon->size(bx); ++i) {
+      const l1t::Muon& mu = muon->at(bx, i);
+
+      //TODO: Still looking for a way to get displaced muon
+      if (abs(mu.hwDXY()) > 0)
+        displaced.push_back(Convertl1tMuon(mu, bx));
+      else
+        prompt.push_back(Convertl1tMuon(mu, bx));
+    }
+
+    // Sort by hwPt
+    std::sort(prompt.begin(), prompt.end(), std::greater<>());
+    std::sort(displaced.begin(), displaced.end(), std::greater<>());
+
+    // Store into output, allow up to 18 prompt + 18 displayed
+    if (prompt.size() > Nprompt) {
+      prompt.resize(Nprompt);
+    }
+    if (displaced.size() > Ndisplaced) {
+      displaced.resize(Ndisplaced);
+    }
+  }
+
+  std::unique_ptr<std::vector<l1t::SAMuon> > prompt_ptr = std::make_unique<std::vector<l1t::SAMuon> >(prompt);
+  std::unique_ptr<std::vector<l1t::SAMuon> > displaced_ptr = std::make_unique<std::vector<l1t::SAMuon> >(displaced);
+  iEvent.put(std::move(prompt_ptr), "promptSAMuons");
+  iEvent.put(std::move(displaced_ptr), "displacedSAMuons");
+}
+
+// ===  FUNCTION  ============================================================
+//         Name:  Phase2L1TGMTSAMuonProducer::Convertl1tMuon
+//  Description:
+// ===========================================================================
+SAMuon Phase2L1TGMTSAMuonProducer::Convertl1tMuon(const l1t::Muon& mu, const int bx_) {
+  ap_uint<BITSQUALITY> qual = mu.hwQual();
+  // We might not need to send BX info
+  //ap_int<BITSBX> bx(bx_);
+  bool charge(mu.charge() > 0);
+  ap_uint<BITSPT> pt = round(mu.pt() / 0.025);
+  ap_int<BITSPHI> phi = round(mu.phi() * (1 << (BITSPHI - 1)) / (M_PI));
+  ap_int<BITSETA> eta = round(mu.eta() * (1 << (BITSETA - 1)) / (M_PI));
+  // FIXME: Below are not well defined in phase1 GMT
+  // Using the version from Correlator for now
+  ap_int<BITSSAZ0> z0 = 0;  // No tracks info in Phase 1
+  ap_int<BITSSAD0> d0 =
+      mu.hwDXY();  // Use 2 bits with LSB = 30cm for BMTF and 25cm for EMTF currently, but subjet to change
+  ap_uint<BITSSABETA> beta = 0;  // No beta from l1t::Muon
+
+  ap_uint<64> word(0);
+  word = word.concat(beta);
+  word = word.concat(d0);
+  word = word.concat(z0);
+  word = word.concat(eta);
+  word = word.concat(phi);
+  word = word.concat(pt);
+  word = word.concat(ap_uint<1>(charge));
+  word = word.concat(qual);
+  //word = word.concat(bx);
```

Comment:
this is different from what we had discussed before and put in the correlator interface document: BITSETA was 12, while here you have 13, and since you've removed the bx all the fields are shifted.

should we update the interface document?

Reply:
also, this piece of code in lines 159-166 doesn't seem to compile, at least in `CMSSW_11_1_7`